javascript/src/gen_package_cg.py

```diff
@@ -0,0 +1,328 @@
+    # ignore folders test and node_modules
+    SOURCE_IGNORE = ["test", "node_modules"]
+    package_dir = os.listdir(package_folder)
+
+    # add the .js files directly in the main directory
+    source_code_in_package = glob.glob(package_folder + "*.js")
+    
+    # loop through all subdirectories and add all .js files there recursively
+    for subdir in filter(lambda x: os.path.isdir(package_folder + x) and x not in SOURCE_IGNORE, package_dir):
+        source_code_in_package.extend(glob.glob(package_folder+subdir+"**/*.js", recursive=True))
+
+    # filter out all paths that contain "test" after the package folder
+    source_code_in_package = list(filter(lambda x: "test" not in x[len(package_folder):], source_code_in_package))
+    return source_code_in_package
+
+ 
+def gen_cg_for_package(package_folder, output_file):
+    """" gen_cg_for_package calculates the call graph for the package in
+    package_folder and its dependencies and saves the call graph to
+    output_file. The call graph is represented as an adjacency list of
+    functions and is saved in json format. 
+
+    Paramaters:
+    package_folder (string): Path to the root package directory containing 
+    package.json and package-lock.json.
+
+    output_file (string): Path to the output file where the call graph
+    is stored. 
+
+    Returns:
+    void
+    """
+    # the rest of the program requires the path to end with / so add it if there isn't any
+    if package_folder[-1] != "/":
+        package_folder += "/"    
+
+    global cg; cg = {}
+    global symbol_ranges_to_footprint; symbol_ranges_to_footprint = {}
+    global interval_trees; interval_trees = {}
+    global incorrect_syntax_files; incorrect_syntax_files = []
+    visited_packages = set()
+
+    # recursive function that will be called for each package in the dep-tree
+    def rec_gen_cg(rec_package_folder):
+
+        # load dependencies
+        try:
+            with open(rec_package_folder + 'package.json') as f:
+                package_conf = json.load(f)
+                dependencies = list(package_conf["dependencies"].keys())
+        except (FileNotFoundError, KeyError) as e:
+            # package.json is not necesary, but in this case we assume the package has no 
+            dependencies = []
+
+        # find all source code
+        source_code_in_curr_package = find_source_files(rec_package_folder)
+
+        # parse all source code files if they have not already been parsed
+        if rec_package_folder not in visited_packages:
+            parse_files(source_code_in_curr_package)
+            visited_packages.add(rec_package_folder)
+
+        # if there are no dependencies we just want to calculate the call graph
+        # for package_folder itself
+        if len(dependencies) == 0:
+            cg_files = source_code_in_curr_package
+            # filter away the files not being parsable
+            cg_files = list(filter(lambda x: x not in incorrect_syntax_files, cg_files))
```

Comment:
Instead of using filter, python list comprehension is a more pythonic way - eg

`cg_files = [x for x in cg_files if x not in incorrect_syntax_files]`

But I'm fine with the lambdas + filter combination, just a headsup.

Reply:
This is also fixed in this commit: https://github.com/debricked/vulnerable-functionality/pull/5/commits/49631838c1408aef23a12328e2058e74d3316fe4